### PR TITLE
added userdefined statistics

### PR DESCRIPTION
**Pull request by MaxOstrowski**
added the option to add userdefined statistics to be set from outside, e.g. clingo scripts

### CODE REVIEW

File: clasp/clasp_facade.h [hunk 1]
```diff
@@ -328,6 +328,76 @@ class ClaspFacade : public ModelHandler {
...
+		/// check whether an Object (of a certain type) is already attached to a map
+		/// if yes, target is set to the objects key
+		bool exists(size_t key, const std::string& name, size_t& target) const;
+		bool exists(size_t key, const std::string& name, Potassco::Statistics_t type, size_t& target) const;
+		/// check whether an Object (of a certain type) is already attached to an array
+		/// if yes, target is set to the objects key
+		bool exists(size_t key, size_t index, size_t& target) const;
+		bool exists(size_t key, size_t index, Potassco::Statistics_t type, size_t& target) const;
+		/// attach a new Object to a Map or Array, or return the old one if if already exists
+		size_t get(size_t key, const std::string& name, Potassco::Statistics_t type);
```
Comment (BenKaufmann) [inline clasp/clasp_facade.h]: Having get() to actually create an object feeels kinda wrong.

File: clasp/clasp_facade.h [hunk 2]
```diff
@@ -328,6 +328,76 @@ class ClaspFacade : public ModelHandler {
...
+		size_t get(size_t key, size_t index, Potassco::Statistics_t type);
+
+		/// (re)sets the value of a value object
+		void set(size_t key, double v);
+
+		StatisticObject toStats();
+private:
```
Comment (BenKaufmann) [inline clasp/clasp_facade.h]: Fix indentation

File: clasp/clasp_facade.h [hunk 3]
```diff
@@ -328,6 +328,76 @@ class ClaspFacade : public ModelHandler {
...
+		void validKey(size_t key) const ;
+		Potassco::Statistics_t type(size_t key) const;
```
Comment (BenKaufmann) [inline clasp/clasp_facade.h]: The private parts should be "pimpl"'ed out into the cpp file.

File: clasp/clasp_facade.h [hunk 4]
```diff
@@ -328,6 +328,76 @@ class ClaspFacade : public ModelHandler {
...
+			size_t target; /// key of the target that is connected here
+		};
+		//! An array of statistic objects.
+		class SimpleStatsVec : public PodVector<StatisticObject>::type {
+		public:
+			StatisticObject toStats() const { return StatisticObject::array(this); }
+		};
+
+		size_t root_;  /// root map index
+		std::unordered_map<size_t, std::vector<Connector>> connections_; /// key -> coneections
```
Comment (BenKaufmann) [inline clasp/clasp_facade.h]: For now please use POTASSCO_EXT_NS::unordered_map

File: clasp/clasp_facade.h [hunk 5]
```diff
@@ -328,6 +328,76 @@ class ClaspFacade : public ModelHandler {
...
+	 */
+	typedef void(*UserdefinedStatsCallback)(UserdefinedStats*, void*);
+	void addUserStatisticsCallback(UserdefinedStatsCallback cb , void* data);
+	//@}
+	//! Gets the callbacks for updating userdefined statistics.
+	/*!
+	 * \param cb returns the callback methods to be called whenever an update is needed.
+	 * \param data returns userdefined callback data.
+	 */
+	void getUserStatisticsCallback(std::vector<UserdefinedStatsCallback>& cbs , std::vector<void*>& data);
```
Comment (BenKaufmann) [inline clasp/clasp_facade.h]: Please be const-correct. Furthermore, I think returning a vector of Closures, where Closure would be a pair<UserdefinedStatsCallback, void*>, would be more natural.

File: clasp/clasp_facade.h [hunk 6]
```diff
@@ -449,6 +519,8 @@ class ClaspFacade : public ModelHandler {
 	BuilderPtr   builder_;
 	SummaryPtr   accu_;
 	StatsPtr     stats_; // statistics: only if requested
+	std::vector<UserdefinedStatsCallback> userStatsCbs_;
```
Comment (BenKaufmann) [inline clasp/clasp_facade.h]: Consider using a single vector storing pairs of function pointer and corresponding data.

File: src/clasp_facade.cpp [hunk 7]
```diff
@@ -555,8 +555,214 @@ double _getResult(const SolveResult* r) { return static_cast<double>(r->operator
...
+
+const char* ClaspFacade::UserdefinedStats::name() const {
+	return "userdefined";
+}
+
+size_t ClaspFacade::UserdefinedStats::root() const {
+	return root_;
+}
+
+bool ClaspFacade::UserdefinedStats::exists(size_t key, const std::string& name, size_t& target) const {
```
Comment (BenKaufmann) [inline src/clasp_facade.cpp]: This is **way** to much duplication!

File: src/clasp_facade.cpp [hunk 8]
```diff
@@ -555,8 +555,214 @@ double _getResult(const SolveResult* r) { return static_cast<double>(r->operator
...
+}
+
+bool ClaspFacade::UserdefinedStats::exists(size_t key, size_t index, size_t& target) const {
+	validKey(key);
+	if (mapper_[key].second != Potassco::Statistics_t::Array) {
+		throw std::logic_error("Statistic function can only be applied to an array");
+	}
+	auto v = connections_.find(key);
+	if (v == connections_.end()) { return false; }
+	for (const auto& i : v->second) {
```
Comment (BenKaufmann) [inline src/clasp_facade.cpp]: Are sparse arrays really needed? The existing stats arrays all behave like normal c++ arrays, i.e. all elements in the range [0, size) are considered valid.

File: src/clasp_facade.cpp [hunk 9]
```diff
@@ -555,8 +555,214 @@ double _getResult(const SolveResult* r) { return static_cast<double>(r->operator
...
+		switch(typeorigin) {
+		case Potassco::Statistics_t::Map: {
+			for (const auto& target : i.second) {
+				maps_[origin].add(strings_[target.index].c_str(), getStats(target.target));
+			}
+			break;
+		}
+		case Potassco::Statistics_t::Array: {
+			size_t size = vecs_[origin].size();
+			/// initialize empty fields with zeros ?
```
Comment (BenKaufmann) [inline src/clasp_facade.cpp]: See comment above. I would already setup the arrays on construction.

File: src/clasp_facade.cpp [hunk 10]
```diff
@@ -555,8 +555,214 @@ double _getResult(const SolveResult* r) { return static_cast<double>(r->operator
...
+			return mapper_[i.target].second == type;
+		}
+	}
+	return false;
+}
+
+size_t ClaspFacade::UserdefinedStats::get(size_t key, const std::string& name, Potassco::Statistics_t type) {
+	size_t target;
+	if (!exists(key,name,type,target)) {
+		switch(type) {
```
Comment (BenKaufmann) [inline src/clasp_facade.cpp]: Please follow clasp's indentation style (this also applies to other places)

File: clasp/clasp_facade.h [hunk 11]
```diff
@@ -328,6 +328,76 @@ class ClaspFacade : public ModelHandler {
 	 */
 	bool               read();
 	//@}
+        
+	/*!
+	 * \name Userdefimed statistics functions
```
[the remark this answers is not in the document]
Comment (pluehne) [reply clasp/clasp_facade.h]: Or, even better, “user-defined.”

**Comment by BenKaufmann** [comment]
Closed. Superseded by https://github.com/potassco/clasp/pull/23

**Pull request closed without merging**